src/Microsoft.Diagnostics.Monitoring/ServerEndpointInfoSource.cs

```diff
@@ -0,0 +1,248 @@
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Diagnostics.NETCore.Client;
+
+namespace Microsoft.Diagnostics.Monitoring
+{
+    /// <summary>
+    /// Aggregates diagnostic endpoints that are established at a transport path via a reversed server.
+    /// </summary>
+    internal class ServerEndpointInfoSource : IEndpointInfoSourceInternal, IAsyncDisposable
+    {
+        // The amount of time to wait when checking if the a endpoint info should be
+        // pruned from the list of endpoint infos. If the runtime doesn't have a viable connection within
+        // this time, it will be pruned from the list.
+        private static readonly TimeSpan PruneWaitForConnectionTimeout = TimeSpan.FromMilliseconds(250);
+
+        private readonly CancellationTokenSource _cancellation = new CancellationTokenSource();
+        private readonly IList<IpcEndpointInfo> _endpointInfos = new List<IpcEndpointInfo>();
+        private readonly SemaphoreSlim _endpointInfosSemaphore = new SemaphoreSlim(1);
+        private readonly string _transportPath;
+
+        private Task _listenTask;
+        private bool _disposed = false;
+        private ReversedDiagnosticsServer _server;
+
+        /// <summary>
+        /// Constructs a <see cref="ServerEndpointInfoSource"/> that aggreates diagnostic endpoints
+        /// from a reversed diagnostics server at path specified by <paramref name="transportPath"/>.
+        /// </summary>
+        /// <param name="transportPath">
+        /// The path of the server endpoint.
+        /// On Windows, this can be a full pipe path or the name without the "\\.\pipe\" prefix.
+        /// On all other systems, this must be the full file path of the socket.
+        /// </param>
+        public ServerEndpointInfoSource(string transportPath)
+        {
+            _transportPath = transportPath;
+        }
+
+        public async ValueTask DisposeAsync()
+        {
+            if (!_disposed)
+            {
+                _cancellation.Cancel();
+
+                if (null != _listenTask)
+                {
+                    await _listenTask.ConfigureAwait(false);
+                }
+
+                _server?.Dispose();
+
+                _endpointInfosSemaphore.Dispose();
+
+                _cancellation.Dispose();
+
+                _disposed = true;
+            }
+        }
+
+        /// <summary>
+        /// Starts listening to the reversed diagnostics server for new connections.
+        /// </summary>
+        public void Listen()
+        {
+            Listen(ReversedDiagnosticsServer.MaxAllowedConnections);
+        }
+
+        /// <summary>
+        /// Starts listening to the reversed diagnostics server for new connections.
+        /// </summary>
+        /// <param name="maxConnections">The maximum number of connections the server will support.</param>
+        public void Listen(int maxConnections)
```

Comment:
It doesn't look like this parameter is used anywhere.  Did you mean to use a different ctor for the `ReversedDiagnosticsServer`?

Reply:
Good catch

Reply:
Fixed